File: apps/web/src/lib/validations/chatbot.ts
import { z } from "zod";

import { TEXT_GENERATION_MODEL_IDS } from "../ai/text-generation/types.ts";

export const chatMessageSchema = z.object({
  content: z.string().trim().min(1).max(8_000),
  role: z.enum(["user", "assistant"]),
});

export const chatbotRequestSchema = z
  .object({
    messages: z.array(chatMessageSchema).min(1).max(24),
    modelId: z.enum(TEXT_GENERATION_MODEL_IDS),
  })
  .refine((request) => request.messages.at(-1)?.role === "user", {
    message: "The last message must be from the user",
    path: ["messages"],
  });

export type ChatMessage = z.infer<typeof chatMessageSchema>;
export type ChatbotRequest = z.infer<typeof chatbotRequestSchema>;

File: apps/ai/app/validations/chatbot.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, model_validator

from app.validations.text_generation import (
    CamelModel,
    GenerateTextResult,
    TextGenerationModelId,
)


class ChatMessage(CamelModel):
    content: str = Field(min_length=1, max_length=8_000)
    role: Literal["user", "assistant"]


class ChatbotRequest(CamelModel):
    messages: list[ChatMessage] = Field(min_length=1, max_length=24)
    model_id: TextGenerationModelId

    @model_validator(mode="after")
    def latest_message_is_from_user(self) -> ChatbotRequest:
        if self.messages[-1].role != "user":
            raise ValueError("The last message must be from the user")
        return self


class ProfileChatbotRequest(ChatbotRequest):
    career_profile: dict[str, object] | None


class ChatbotAIResult(GenerateTextResult):
    requires_career_profile: bool
